### DISCUSSION TITLE
Recommend a better encryption scheme

### DISCUSSION THREAD

**Discussion by Kung-Fu-Master** [minio/minio#12326] [Q&A] [answered]
It's better to use the standard HKDF to run a key-derivation algorithm to generate Object Key.

It's better to use the standard HKDF to run a key-derivation algorithm to generate the KEK(Key Encryption Key). Adding two components (counter, length) in key derivation is a well-established crypto BKM for KDFs ensuring robust domain separation for keys and it will help to reliably (even if RNG fails/under attacker's control) separate keys if/when MinIO needs to derive several keys of same/different lengths for a given bucket/path.

HKDF reference Link: https://tools.ietf.org/html/rfc5869

**Comment by aead** [comment, marked as answer]
Currently, there are three distinct cryptographic keys for a given SSE-encrypted S3 object.
 - The "external" key `EK` is provided by the client or a KMS.

 - The object encryption key `OEK` is *always* 256 bits long and generated at random. More specifically it is computed as 
    `SHA-256(EK, nonce)` where `nonce` is a 256 bit random value. The purpose of computing a `SHA-256` hash here is
    to ensure that the `OEK` is unpredictable even if the (P)RNG fails completely under the assumption that the `EK` is 
    unpredictable. As a side-effect of SHA-256 the computed `OEK` is always 256 bits long. The purpose of the `nonce`
    here is to ensure that given the same `EK` for two objects results in two different unpredictable `OEK`s assuming the
   (P)RNG is at least non-repeating - not necessarily indistinguishable from true randomness. The only assumption here
   that `SHA-256` is collision-resistant. As of today this is assumed to be true.

 - The key encryption key `KEK` is *always* 256 bits long. It is used to encrypt the `OEK` and is derived from the `EK` 
    using the following scheme:
   ```
   HMAC-SHA256(EK, IV || DOMAIN || ALGORITHM || BUCKET || "/" || OBJECT)
   ```
    Here, the `IV` is a 256 bit random value, `DOMAIN` is one of the following strings depending on the SSE type:
   `"SSE-S3"`, `"SSE-C"`, `"SSE-KMS"`, `ALGORITHM` is a string naming the encryption scheme used for the object itself.
    Finally, `BUCKET` is the bucket name and `OBJECT` is the object name joined together.
   
    The purpose of the `DOMAIN` string is to fix the SSE type. Trying to change the SSE type later on will result
    in a different `KEK` (see assumptions later on). The purpose of the `ALGORITHM` string is to fix the object 
    encryption scheme, and therefore, analogous to the `DOMAIN`. The purpose of `BUCKET` and `OBJECT`  is 
    to fix the specific object path and prevent renaming / moving the object. Doing so results in a different object path
    (path names bijective), and therefore, results in a different `KEK` (again see assumptions later on).
    Finally, the `IV` is a random value to make the key derivation probabilistic and ensure that you derive unpredictable 
   `KEK`s given the same `EK` assuming the (P)RNG is at least non-repeating.

    This key derivation scheme relies on the following security assumption: *HMAC-SHA256 is a PRF.* This is formally 
    proven assuming `SHA-256` is pre-image  collision-resistant. As of today this is assumed to be true.
    However, this scheme **can only** derive 256 bit keys.

***

`HKDF` has the following advantages compared to the `HMAC-SHA256` scheme described above:
  - Key expansion: `HKDF` can generate (more or less) arbitrary long and independent cryptographic keys - not just 256 bit 
     long keys - due to the `expand` construction.

However, this difference between `HKDF` and the `HMAC-SHA256` scheme shown above is not relevant in the specific case here since MinIO does not derive keys of variable lengths.

It's worth noting that the `HMAC-SHA256` construction basically falls under:
> In the case where the amount of required key bits, L, is no more than
   HashLen, one could use PRK directly as the OKM.  This, however, is
   NOT RECOMMENDED, especially because it would omit the use of 'info'
   as part of the derivation process (and adding 'info' as an input to
   the extract step is not advisable -- see [HKDF-paper]).

> https://datatracker.ietf.org/doc/html/rfc5869

However, the problem of "traditional KDFs" mentioned in Sec 8 of https://eprint.iacr.org/2010/264.pdf explicitly does not apply to the `HMAC-SHA256` construction above. 
 

***

In any case future versions of MinIO may use a more optimised and eventually better suited key derivation mechanism - eventually based on sponge constructions.
